src/ProtocolGateway.Core/Mqtt/MessageBatchAsyncProcessor.cs

```diff
@@ -0,0 +1,148 @@
+        readonly Queue<PublishPacket> backlogQueue;
+        State state;
+        readonly TaskCompletionSource closedPromise;
+
+        public MessageBatchAsyncProcessor(int maxBatchSizeInBytes, Func<IChannelHandlerContext, IReadOnlyCollection<PublishPacket>, Task> processFunc)
+        {
+            this.maxBatchSizeInBytes = maxBatchSizeInBytes;
+            this.processFunc = processFunc;
+            this.backlogQueue = new Queue<PublishPacket>();
+            this.closedPromise = new TaskCompletionSource();
+        }
+
+        public Task Closed => this.closedPromise.Task;
+
+        public int BacklogSize => this.backlogQueue.Count;
+
+        public void Post(IChannelHandlerContext context, PublishPacket packet)
+        {
+            switch (this.state)
+            {
+                case State.Idle:
+                    this.backlogQueue.Enqueue(packet);
+                    this.state = State.Processing;
+                    this.StartQueueProcessingAsync(context);
+                    break;
+                case State.Processing:
+                    this.backlogQueue.Enqueue(packet);
+                    break;
+                case State.Closed:
+                    ReferenceCountUtil.Release(packet);
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException();
+            }
+        }
+
+        public void Close()
+        {
+            switch (this.state)
+            {
+                case State.Idle:
+                    this.state = State.Closed;
+                    this.closedPromise.TryComplete();
+                    break;
+                case State.Processing:
+                    this.state = State.Closed;
+                    Queue<PublishPacket> queue = this.backlogQueue;
+                    while (queue.Count > 0)
+                    {
+                        PublishPacket packet = queue.Dequeue();
+                        ReferenceCountUtil.SafeRelease(packet);
+                    }
+                    break;
+                case State.Closed:
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException();
+            }
+        }
+
+        async void StartQueueProcessingAsync(IChannelHandlerContext context)
+        {
+            try
+            {
+                Queue<PublishPacket> queue = this.backlogQueue;
+                while (queue.Count > 0 && this.state != State.Closed)
+                {
+                    int batchSize = 0;
+                    List<PublishPacket> messages = new List<PublishPacket>();
+                    do
+                    {
+                        PublishPacket message = queue.Peek();
+                        int messageSize = message.Payload?.ReadableBytes ?? 0 + message.TopicName.Length;
+                        if (messageSize + batchSize <= this.maxBatchSizeInBytes)
+                        {
+                            queue.Dequeue();
```

Comment:
Shouldn't you dequeue after ProcessAsync ?

Reply:
... I guess you should not.

Reply:
we can't retry here and failure would bring whole connection down, so not a concern